Upgrade nbdime for jupyterlab 4

HaudinFlorence opened the pull request:
This PR is a first pass on trying a migration to jupyterlab4. It includes :
- moving from Lumino 1 to 2
- moving from JupyterLab 2 || 3 to 4 working with codemirror 6 (CM6)

**Main changes**
This PR includes among others the migration to CM6 of these features of the web applications:
- lines and characters hilighting using a dedicated a stateField, stateEffects and specific decorations of CM6
- spacers when addition/deletion of texts are made using a stateField, stateEffects, a builder and decoration widgets of CM6
- picker and conflict gutter markers using dedicated stateFields and stateEffects
- update of the different methods that enable to update the views of the editors and the models of the diffView.

**Diff app rendering example**
![comparisons_diff_example8](https://github.com/jupyter/nbdime/assets/99649086/4baad8e6-63d0-4acb-bc7e-26d6abfc1c6a)

**Merge app rendering example**
![comparisons_merge_example8](https://github.com/jupyter/nbdime/assets/99649086/3cc28b13-c686-403e-94c2-988505e65df9)

**Styling/UX differences**
- the collapse functionnality has been removed.
- the new background is white when it was previously light grey
- in the merge app, the padding spacers don't have colors. They are ll grey (using decoration widgets makes the previous css logics not work anymore, this should be changed to make the coloring possible like in the diff web application.)
- in the diff app, the small arrows between editors have been removed. There are picker gutter markers in the diff too.
- all padding spacers are directly displayed on the view, there is no need to click on a picker gutter markers to make a spacer fully expend with its real sizing.
- there is no picker gutter marker for spacers. In the merge editor, only the text chunks are shown, there is no spacer.
- the syntax highlighting seems to have priority on the character change highlighting (as shown for the `sinus` function in the central editor in the merge example above).
- there is only 1 gutter when there were 2 previously (1 for pickers and 1 for the conflict gutter markers)
- there is a small white space between editors defined in the css of the grid panel now used to encapsulate the editors.
- 'Metadata changed' panel at the end of the applications is missing.

 **Future work and follow up tasks** 
- In some cases, the result of the native function `findAlignedLines` is not right leading to a misalignment of lines to be put at the same height and to incorrect treatment of adding paddings. An issue has been opened with an example #669

- More ui tests will be added when `findAlignedLines` will be fixed for the non working case

- In reference to comment https://github.com/jupyter/nbdime/pull/664#discussion_r1314273776, changes will have to be made after the PR is merged regarding use of  `find`, `each`, `toArray` related to Lumino to move to JS equivalent.

vidartf commented on .gitignore:
```diff
@@ -12,6 +12,7 @@ MANIFEST
 .cache
 .pytest_cache
 .idea
+.yarn
```
Note that this repo is currently using npm + package-lock.json, so please avoid adding yarn things.

HaudinFlorence replied:
@vidartf Ok. I will change this.

vidartf commented on package.json:
```diff
@@ -1,21 +1,24 @@
-    "lerna": "^4.0.0",
-    "rimraf": "^5.0.0"
-  }
+	"private": true,
```
There are a lot of formatting changes in this PR. It would be very much easier for me to look at it if pure formatting changes were removed! 😃

HaudinFlorence replied:
@vidartf That's right. Would you be ok to add a formatter like prettier on the project? This could be done in a separate PR. If so I can do it. The changes on migration would then be appear more clearly. Please let me know what you think about it ?

HaudinFlorence replied:
@vidartf Thanks for your reply. So we can try to do what you are suggesting : " a config for prettier + an autoformat commit + the ignore config so that GH doesn't consider that commit for git blame." As a first step, I opened this draft PR https://github.com/jupyter/nbdime/pull/667 with the config for prettier (that can be tuned), and a commit with the autoformat changes. I guess that we can add in a second time a `.git-blame-ignore-revs` so that Github does not consider this commit for `git blame`. Let me please know if the prettier config does not fit your expectations.

JohanMabille commented on packages/nbdime/src/common/mergeview.ts:
```diff
@@ -143,563 +341,404 @@ function createNbdimeMergeView(
+    );
+
+    this.syncScroll(baseEditor, remoteEditor);
+    this.syncScroll(remoteEditor, baseEditor);
```
I think these functions should be called in init instead, no need to add a new listener each time the view updates.

HaudinFlorence replied:
Thanks. This has been changed.

JohanMabille commented on packages/nbdime/src/common/mergeview.ts:
```diff
@@ -143,563 +341,404 @@ function createNbdimeMergeView(
-  syncModel() {
-    if (this.modelInvalid()) {
-      let edit = this.ownEditor;
+    syncModel() {
```
Formatting: Too much indentation for the whole function

HaudinFlorence replied:
Thanks. This has been corrected.

JohanMabille commented on packages/nbdime/src/common/mergeview.ts:
```diff
@@ -143,563 +341,404 @@ function createNbdimeMergeView(
+      editor.dispatch({
+        changes: {from: 0, to: editor.state.doc.length, insert: newLines.slice(0, newLines.length).join('')}
+      });
+      this.remoteEditorWidget.cm.dispatch({selection: {anchor: cursor}})
```
Nitpicking: use editor instead.
I think you can dispatch the changes and the selection in a single call

HaudinFlorence replied:
Thanks. This has been taken into account.

JohanMabille commented on packages/nbdime/src/common/mergeview.ts:
```diff
@@ -143,563 +341,404 @@ function createNbdimeMergeView(
     this.lockScroll = val;
     if (val && action !== false) {
-      this.syncScroll(EventDirection.OUTGOING);
+      this.syncScroll(this.baseEditorWidget.cm, this.remoteEditorWidget.cm);
```
I don't htink we need this call here, scroll synchronization should be done by listeners registered in the init method.

HaudinFlorence replied:
Thanks. This should have been taken into account

JohanMabille commented on packages/nbdime/src/common/mergeview.ts:
```diff
@@ -143,563 +341,404 @@ function createNbdimeMergeView(
    * Sync scrolling between base and own editors. `type` is used to indicate
    * which editor is the source, and which editor is the destination of the sync.
    */
-  protected syncScroll(type: EventDirection): void {
```
- The comment should be updated to reflect the changes
- The method should be private

HaudinFlorence replied:
Thanks. This has been modified.

JohanMabille commented on packages/nbdime/src/common/mergeview.ts:
```diff
@@ -143,563 +341,404 @@ function createNbdimeMergeView(
+    let destScroller = destEditor.scrollDOM;
+    srcScroller.addEventListener("scroll", (event) => {
+      window.requestAnimationFrame(function() {destScroller.scrollLeft = srcScroller.scrollLeft;})
     });
```
It would be more natural to have this function add the eventListener to both editors instead of having to call it twice.

HaudinFlorence replied:
Thanks. This has been taken into account.

JohanMabille commented on packages/nbdime/src/common/mergeview.ts:
```diff
@@ -143,563 +341,404 @@ function createNbdimeMergeView(
+    chunk: chunk,
+    type: type
+  });
+  effects.push(mappingEffect);
```
Nitpicking: this function can be shorten by removing all the local variables (except effects), for instance:
```ts
effects.push(addLineChunkMappingEffect.of({
  line: offsetToPos(editor.state.doc, pos).line,
  chunk: chunk,
  type: type
}));
```

HaudinFlorence replied:
Thanks. This has been taken into account.

JohanMabille commented on packages/nbdime/src/common/mergeview.ts:
```diff
@@ -143,563 +341,404 @@ function createNbdimeMergeView(
+      effects.push(removeHighlightEffect.of({highlightType: 'chunk', decorationKey: decorationKey}));
+      effects.push(removeHighlightEffect.of({highlightType: 'conflict', decorationKey: decorationKey}));
+      effects.push(removeHighlightEffect.of({highlightType: 'start', decorationKey: decorationKey}));
+      effects.push(removeHighlightEffect.of({highlightType: 'end', decorationKey: decorationKey}));
```
This can be shorten with a `for ... of` loop

HaudinFlorence replied:
Thanks. This has been changed.

JohanMabille commented on packages/nbdime/src/common/mergeview.ts:
```diff
@@ -143,563 +341,404 @@ function createNbdimeMergeView(
-      classes[k] += postfix;
+    for (let chunk of chunkArray) {
+      let sources: ChunkSource[] = chunk.sources;
+      let decorationKey = this.getDecorationKey(sources);
```
You can use directly `chunk.sources` here to avoid introducing a local variable

HaudinFlorence replied:
Thanks. This has been done.

JohanMabille commented on packages/nbdime/src/common/mergeview.ts:
```diff
@@ -143,563 +341,404 @@ function createNbdimeMergeView(
+        if (r.source !== undefined) {
+          sources.push(r.source);
+        }
+        let decorationKey = this.getDecorationKey(sources);
```
No need for this variable, you can initialize the `addHighlightEffect` with a call to `getDecorationKey`

HaudinFlorence replied:
Thanks. This has been changed.

JohanMabille commented on packages/nbdime/src/common/mergeview.ts:
```diff
@@ -143,563 +341,404 @@ function createNbdimeMergeView(
+          highlightType: highlightType,
+          decorationKey: decorationKey
+        }));
       }
```
All the local variables of this function should be const

HaudinFlorence replied:
Thanks. This has been modified.

JohanMabille commented on packages/nbdime/src/common/mergeview.ts:
```diff
@@ -143,563 +341,404 @@ function createNbdimeMergeView(
+          sources.push(r.source);
+        }
+        let decorationKey = this.getDecorationKey(sources);
+        effects.push(removeHighlightEffect.of({highlightType: highlightType, decorationKey: decorationKey}));
```
Use `getDecorationKey` when initializing the effect, to avoid introducing a local variable

HaudinFlorence replied:
Thanks. The change has been done.

JohanMabille commented on packages/nbdime/src/common/mergeview.ts:
```diff
@@ -143,563 +341,404 @@ function createNbdimeMergeView(
+    markType: DIFF_OP
+  ) {
+    let effects: StateEffect<unknown>[] = [];
+    let sources: ChunkSource[] = [];
```
These variables should be const

HaudinFlorence replied:
Thanks. These 2 variables have been set to const.

JohanMabille commented on packages/nbdime/src/common/mergeview.ts:
```diff
@@ -143,563 +341,404 @@ function createNbdimeMergeView(
-
 
-// Updating the gap between editor and original
+  protected updateHighlighting(
```
This method should be private

HaudinFlorence replied:
Thanks. This has been done.

JohanMabille commented on packages/nbdime/src/common/mergeview.ts:
```diff
@@ -143,563 +341,404 @@ function createNbdimeMergeView(
+    chunkArray: Chunk[],
+    type: DIFF_OP
+  ) {
+    let self = this;
```
No need to do that

HaudinFlorence replied:
Thanks. The definition and use of `self` has been removed.

HaudinFlorence replied:
Thanks. The definition and use of `self` has been removed.

JohanMabille commented on packages/nbdime/src/common/mergeview.ts:
```diff
@@ -143,563 +341,404 @@ function createNbdimeMergeView(
+      self.clearCharacterHighlighting(editor, diffRanges, type);
+    let effects: StateEffect<unknown>[] =
+      clearLineEffects.concat(clearCharacterHighlightEffects);
+    editor.dispatch({ effects });
```
Same remarks here:
- variables should be const
- no need for `self=this`
- the method should be private

HaudinFlorence replied:
This has been done.

JohanMabille commented on packages/nbdime/src/common/mergeview.ts:
```diff
@@ -143,563 +341,404 @@ function createNbdimeMergeView(
+  lineChunks: Chunk[];
+  gap: HTMLElement;
+  lockScroll: boolean;
+  protected lockButton: HTMLElement;
```
All the data members should be private

HaudinFlorence replied:
Thanks. This has been corrected

JohanMabille commented on packages/nbdime/src/common/mergeview.ts:
```diff
@@ -945,311 +968,292 @@ class MergeView extends Panel {
-          return function() { dv.ownEditor.operation(fn); };
-        }(f);
+
+  protected onGutterClick(editor: EditorView, line: BlockInfo): boolean {
```
This should be private

HaudinFlorence replied:
This has been changed too.

JohanMabille commented on packages/nbdime/src/common/mergeview.ts:
```diff
@@ -945,311 +968,292 @@ class MergeView extends Panel {
   merge: DiffView | null;
   base: EditorWidget;
   options: any;
   diffViews: DiffView[];
```
All the data members should be private

HaudinFlorence replied:
Thanks. This has been modified?

JohanMabille reviewed (commented):
I think it would be nice to add documentation that shortly describes the new fields / effects and how they interact / are used in the diff view and the merge view.

HaudinFlorence commented:
> I think it would be nice to add documentation that shortly describes the new fields / effects and how they interact / are used in the diff view and the merge view.

@JohanMabille Thanks for the review. The documentation will be added in a following commit.

HaudinFlorence commented on packages/nbdime/src/common/mergeview.ts:
```diff
@@ -143,563 +341,404 @@ function createNbdimeMergeView(
+      self.buildCharacterHighlighting(editor, diffRanges, type);
+    let effects: StateEffect<unknown>[] =
+      LineHighlightEffects.concat(MarkHighlightEffects);
+    editor.dispatch({ effects });
```
Thanks. This has been changed.

HaudinFlorence commented:
This PR is now ready for review. Feel free to tag or propose other potential reviewers. Thanks.
@vidartf @krassowski

krassowski reviewed (changes requested):
Thank you so much for working on it! First review pass:
- after compiling locally and clicking the diff button in JupyterLab I get a `TypeError` which appears to be a problem with porting iterator (details in a comment on `stringified.ts`)
- I wonder if some more CSS rules should be removed -  but I was not able to test this PR further due to the above; I added comments in some places for myself to check later - please disregard comments if you have already looked at these
- there is number of commented code lines that should be removed
- some functionality appears to be removed - at least from scrolling the diff, or maybe offered by upstream (?) - can you add a high level overview of these changes in the PR description to assist with review and for future reference?
- thank you so much for adding playwright tests!
   - there are playwright report files in the PR, including videos; please remove them and add to gitignore
   - do you plan to make the playwright tests run on CI? If you already have a `.github/` workflow, it does not show up in the PR
   - I see files for test1 and test3 but not for test2, is this intended?
- there may be some changes to `setup.py` to make, such as including relevant classifiers (see extension cookiecutter) so that package is picked up by JupyterLab extension manager, and checking that all dependencies and version ranges are still right.

krassowski commented on ui-tests/tests/jupyter_theme_editor.spec.ts:
```diff
@@ -0,0 +1,27 @@
+});
+
+test('choose left solution for conflict',  async ({ page }) => {
+
```
Also, the name of this test file appears to be irrelevant?

HaudinFlorence replied:
This file was a leftover. Sorry for that. It has been removed.

HaudinFlorence commented:
> Thank you so much for working on it! First review pass:
> 
> * after compiling locally and clicking the diff button in JupyterLab I get a `TypeError` which appears to be a problem with porting iterator (details in a comment on `stringified.ts`)
> * I wonder if some more CSS rules should be removed -  but I was not able to test this PR further due to the above; I added comments in some places for myself to check later - please disregard comments if you have already looked at these
> * there is number of commented code lines that should be removed
> * some functionality appears to be removed - at least from scrolling the diff, or maybe offered by upstream (?) - can you add a high level overview of these changes in the PR description to assist with review and for future reference?
> * thank you so much for adding playwright tests!
>   
>   * there are playwright report files in the PR, including videos; please remove them and add to gitignore
>   * do you plan to make the playwright tests run on CI? If you already have a `.github/` workflow, it does not show up in the PR
>   * I see files for test1 and test3 but not for test2, is this intended?
> * there may be some changes to `setup.py` to make, such as including relevant classifiers (see extension cookiecutter) so that package is picked up by JupyterLab extension manager, and checking that all dependencies and version ranges are still right.

@krassowski Thanks a lot for the review. I will try to take into account your comments in a next commit

HaudinFlorence commented on packages/nbdime/src/common/mergeview.ts:
```diff
@@ -921,335 +1018,349 @@ class MergeView extends Panel {
   }
-}
 
-function collapseIdenticalStretches(mv: MergeView, margin?: boolean | number): void {
```
Yes this has been removed. I have added a line in the description of the PR to inform about it.

HaudinFlorence commented on packages/nbdime/src/merge/widget/cell.ts:
```diff
@@ -126,14 +126,16 @@ class CellMergeWidget extends Panel {
   validateMerged(candidate: nbformat.ICell): nbformat.ICell {
     if (this.sourceView && this.sourceView instanceof MergeView) {
-      let text = this.sourceView.getMergedValue();
+      /*let text = this.sourceView.getMergedValue();*/
+      let text = '';
```
I don't know how to change this.

HaudinFlorence commented on packages/nbdime/src/diff/range.ts:
```diff
@@ -65,8 +67,10 @@ class DiffRangePos {
   constructor(
-        public from: CodeMirror.Position,
-        public to: CodeMirror.Position,
+        //public from: CodeMirror.Position,
+        //public to: CodeMirror.Position,
```
Thanks. This comment has been removed

HaudinFlorence commented on packages/nbdime/src/diff/range.ts:
```diff
@@ -132,17 +144,17 @@ function raw2Pos(raws: DiffRangeRaw[], text: string): DiffRangePos[] {
     let lineStartIdx = line > 0 ? adIdx[line - 1] + 1 : 0;
-    let from = CodeMirror.Pos(line, r.from - lineStartIdx);
-
+    //let from = CodeMirror.Pos(line, r.from - lineStartIdx);
```
Thanks. This comment has been removed

HaudinFlorence commented on packages/nbdime/src/patch/common.ts:
```diff
@@ -6,9 +6,9 @@ import type {
-import type {
-  IIterator
-} from '@lumino/algorithm';
+/* import type {
```
Thanks. This comment has been removed

HaudinFlorence commented on ui-tests/playwright-report/data/f82b41ea230488c9e9fbb81d60195e34f46aa484.png:
Thanks. This should normally be taken into account in the next commit.

HaudinFlorence commented on packages/webapp/package.json:
```diff
@@ -12,18 +12,18 @@
+    "@jupyterlab/cells": "^4.0.0",
+    "@jupyterlab/codemirror": "^4.0.0",
+    "@jupyterlab/coreutils": "^6.0.0",
+    "@jupyterlab/mathjax-extension": "^4.0.0",
```
This will be removed.

HaudinFlorence commented on packages/labextension/package.json:
```diff
@@ -42,26 +42,26 @@
+    "@lumino/coreutils": "^2.0.0",
+    "@lumino/disposable": "^2.0.0",
+    "@lumino/widgets": "^2.0.0",
+    "nbdime": "^6.1.2"
```
If I have put nbdime version to 6.2.0 in this package.json file. I have also put lumino packages to the latest versions in the `package.json` of `/packages/nbdime` and of `/packages/labextension`

HaudinFlorence commented on packages/labextension/package.json:
```diff
@@ -42,26 +42,26 @@
+    "@jupyterlab/builder": "^4.0.0",
+    "@jupyterlab/docregistry": "^4.0.0",
+    "@lumino/commands": "^2.0.0",
+    "mkdirp": "^0.5.1",
+    "rimraf": "^2.6.3",
```
This is a mistake. This has been restablished to versions used in Nbdime 6.2.

HaudinFlorence commented on packages/nbdime/src/common/editor.ts:
```diff
@@ -3,67 +3,108 @@
-        return factory.newInlineEditor.bind(factory);
-      }()
+    )) {
+      if (extensionFactory.name === 'lineNumbers') {
+        extensions.addExtension(extensionFactory);
+      }
+      if (extensionFactory.name === 'readOnly') {
+        extensions.addExtension(extensionFactory);
+      }
+      if (extensionFactory.name === 'theme') {
+        extensions.addExtension(extensionFactory)
+      }
+      if (extensionFactory.name === 'allowMultipleSelections') {
+        extensions.addExtension(extensionFactory);
+      }
+      if (extensionFactory.name === 'tabSize') {
+        extensions.addExtension(extensionFactory);
+      }
```
Thanks. This has been changed.

HaudinFlorence commented on packages/nbdime/src/styles/diff.css:
```diff
@@ -126,38 +124,51 @@
+.jp-Notebook-diff .cm-right-editor .cm-merge-spacer { background-color: var(--jp-diff-added-color2); }
 
+.jp-Notebook-diff .cm-left-editor .cm-line .cm-merge-r-deleted {
+  /*background-color: initial;*/
```
This has been changed.

HaudinFlorence commented on packages/webapp/src/app/staticdiff.ts:
```diff
@@ -30,43 +30,42 @@ for (var i=0;i<headers.length;++i){
 
 function ensureRendered(callback: () => void): void {
-  for (let e of EditorWidget.editors) {
+  /*for (let e of EditorWidget.editors) {
```
Thanks. This has been commented out.

HaudinFlorence commented on packages/nbdime/src/styles/common.css:
```diff
@@ -1,10 +1,10 @@
 
+
 .nbdime-root .CodeMirror-merge {
```
I don't think so. I have removed this styles.

HaudinFlorence commented on packages/nbdime/src/legacy_codemirror/cmconfig.ts:
```diff
@@ -0,0 +1,16 @@
+
+export namespace LegacyCodeMirror {
+    export interface EditorConfiguration {
```
This has been modified.

HaudinFlorence commented on packages/nbdime/src/legacy_codemirror/cmconfig.ts:
```diff
@@ -0,0 +1,16 @@
+        /** Whether to show line numbers to the left of the editor. */
+        lineNumbers?: boolean;
+
+        /** This disables editing of the editor content by the user. If the special value "nocursor" is given (instead of simply true), focusing of the editor is also disallowed. */
```
I am sorry but I don't understand what you mean.

krassowski replied:
```suggestion
        /** This disables editing of the editor content by the user. */
```

because the interface specifies `readOnly?: boolean`

HaudinFlorence replied:
Sorry for the confusion. I can propose `boolean | string`

HaudinFlorence commented on packages/nbdime/src/patch/stringified.ts:
```diff
@@ -210,7 +205,7 @@ function patchStringifiedObject(base: ReadonlyJSONObject, diff: IDiffObjectEntry
   let helper = new PatchObjectHelper(base, diff);
   let baseKeys = helper.baseKeys.slice();
-  each(helper.keys(), key => {
+  for( const key in helper.keys()) {
```
I am not sure about how to fix that.

HaudinFlorence commented on ui-tests/tests/nbdime-merge-test3.spec.ts:
```diff
@@ -0,0 +1,34 @@
+
+
+
+/*Test3: Notebooks of same length 0 conflict */
```
If correctly implemented, this should has been taken into account.

HaudinFlorence commented on packages/webapp/src/app/common.ts:
```diff
@@ -18,23 +18,23 @@ import type {
 import {
   NotifyUserError
-} from 'nbdime/lib/common/exceptions';
+} from '../../../nbdime/lib/common/exceptions';
```
This is a wrong leftover. Sorry for that. This will be removed.

HaudinFlorence commented on packages/nbdime/src/styles/merge.css:
```diff
@@ -81,85 +74,138 @@
+.jp-Notebook-merge .cm-central-editor .cm-line .cm-merge-r-deleted {
   background-color: initial;
   color: #b00
 }
```
They have been removed. There still remains:
- in `diff.css`
* `.jp-Notebook-diff .jp-Diff-deleted .CodeMirror,
.jp-Notebook-diff .jp-Diff-deleted .jp-Diff-renderedOutput,
.jp-Cellrow-outputs .jp-Diff-twoway .jp-Diff-base`
- `.jp-Notebook-diff .jp-Diff-added .CodeMirror,
.jp-Notebook-diff .jp-Diff-added .jp-Diff-renderedOutput,
.jp-Cellrow-outputs .jp-Diff-twoway .jp-Diff-remote`
- in merge.css
* `.jp-Notebook-merge .CodeMirror pre.cm-line {
  z-index: 3;
}`

that I think should be removed to.
-

HaudinFlorence commented:
Hello@vidartf Do you think it would be possible to approve the workflow? Thanks.

krassowski commented:
> I am not sure to understand the error you are encountering. Is it when you try to use one the the 2 apps nbmerge-web or nbdiff-web?

No, this is using JupyterLab extension shipped by nbdime.

![demon-nbdime](https://github.com/jupyter/nbdime/assets/5832902/5a80120b-3286-4a00-938b-f3b1199576c7)

krassowski commented:
Just to stress, this is not `jupyterlab-git`, but core nbdime functionality implemented here: [jupyter/nbdime/tree/master/packages/labextension](https://github.com/jupyter/nbdime/tree/master/packages/labextension).

krassowski commented:
Which exact version of JupyterLab do you have? How do you build the extension? I pulled latest changes, run:

```bash
rm -r nbdime/labextension/
npm install
npm run build
jupyter lab
```

The build stop succeeded with:

```
webpack 5.86.0 compiled with 10 warnings in 42427 ms
lerna success run Ran npm script 'build' in 3 packages in 50.4s:
lerna success - nbdime-jupyterlab
lerna success - nbdime
lerna success - nbdime-webapp
```

and I am still getting the same error.

krassowski commented:
I can reproduce the `Test / JavaScript` failures locally. Adding types for jest does not solve it, it looks like more work is needed to port the tests.

HaudinFlorence commented:
> Which exact version of JupyterLab do you have? How do you build the extension? I pulled latest changes, run:
I am using 4.0.0b1. 
For building, I am running `npm run build.` and then`nbdime extensions --enable [--sys-prefix/--user/--system]`.

HaudinFlorence commented:
> I can reproduce the `Test / JavaScript` failures locally. Adding types for jest does not solve it, it looks like more work is needed to port the tests.

I haven't really tried yet to make the tests work and I have almost zero experience in writing new ones.

vidartf commented:
> Hi @vidartf. Do you think it would be possible to please approve the workflow? Thanks.

I'm currently traveling, so I will review once I get back (next week), but it looks like someone else approved the workflow for now?

krassowski commented:
Hi @vidartf, yes I reached out to @Carreau who approved the run and I will keep hitting the button if needed to keep this one moving.

jtpio commented:
> I can reproduce the `Test / JavaScript` failures locally. Adding types for jest does not solve it, it looks like more work is needed to port the tests.

Maybe there are more updates to do based on https://jupyterlab.readthedocs.io/en/latest/extension/extension_migration.html#testing-with-jest.

jtpio commented:
Adding `@types/jest` seems to be helping with running the tests: https://github.com/HaudinFlorence/nbdime/pull/2

Some are failing though, but at least they run:

![image](https://github.com/jupyter/nbdime/assets/591645/6294e212-8363-4a58-99a5-498f07a17d16)

krassowski commented on setup.py:
```diff
@@ -128,7 +128,6 @@
     'requests',
     'GitPython!=2.1.4, !=2.1.5, !=2.1.6',  # For difftool taking git refs
     'jupyter_server',
-    'jupyter_server_mathjax>=0.2.2',
```
I suggested considering this because MathJax4 implementation in JupyterLab 4 now ships all static assets by itself, so there is no need for it. Indeed The import in the server, `from jupyter_server_mathjax.app import STATIC_ASSETS_PATH` and related code would need to be removed too.

jtpio replied:
Ah ok then let's clean that up :+1:

HaudinFlorence replied:
@jtpio @krassowski Thanks for your comments. I will clean up and push.

HaudinFlorence commented:
> Hi @vidartf, yes I reached out to @Carreau who approved the run and I will keep hitting the button if needed to keep this one moving.

Thanks @vidartf @Carreau @krassowski Isn't there a way to make it possible for once ?

krassowski commented:
One neat trick I use is to open a PR against your own repository (after enabling CI in its settings) - this way you can see build status even without having previously contributed to a repository.

HaudinFlorence commented:
@krassowski I could not fix properly the iterator issue. If you have a solution, I will be happy to include it.

krassowski commented:
> @krassowski I could not fix properly the iterator issue for the moment. If you have a solution, I will be happy to include it.

@HaudinFlorence please see https://github.com/HaudinFlorence/nbdime/pull/4

HaudinFlorence commented:
> > @krassowski I could not fix properly the iterator issue for the moment. If you have a solution, I will be happy to include it.
> 
> @HaudinFlorence please see [HaudinFlorence#4](https://github.com/HaudinFlorence/nbdime/pull/4)

Thanks a lot Mike @krassowski I have merged your PR. As you mentionned, JS tests are now ok, there is still one python tets failing. If someone could please approve the workflow? Thanks.

krassowski commented:
@HaudinFlorence please see a fix for the failing Python test over in https://github.com/HaudinFlorence/nbdime/pull/7 :)

> Note to self: we probably should add a similar line to unit tests in JupyterLab itself, it can give quite a boost.

HaudinFlorence commented:
> @HaudinFlorence please see a fix for the failing Python test over in [HaudinFlorence#7](https://github.com/HaudinFlorence/nbdime/pull/7) :)
> 
> > Note to self: we probably should add a similar line to unit tests in JupyterLab itself, it can give quite a boost.

Thanks @krassowski I have added the fix.

HaudinFlorence commented:
I have merged a PR from my fork that adds 2 ui-test to the github workflow (more will be added in the future). The windows tests are failing because the reference snapshots seemed not to be produced. For the linux ones, I first produced them locally on my machine and pushed them, so they still exists when running the test for the first time. In any case, I thought that first test run was supposed to produce the reference snapshots and second run would be ok for comparison between the reference and the generated one. This doesn't seem to be the case. Any hint for this is please welcomed (sorry I am quite new with implementing ui-tests).

krassowski commented:
You can reuse [`update-integration-tests.yml`](https://github.com/jupyterlab/extension-template/blob/e7e4c2bd3467aa3c5a8864e438e897c5f0439098/template/.github/workflows/%7B%25%20if%20test%20%25%7Dupdate-integration-tests.yml%7B%25%20endif%20%25%7D) from the extension template for updating snapshots. We would then use comments with `please update playwright snapshots` text to trigger that workflow.

krassowski commented on .github/workflows/tests.yml:
```diff
@@ -28,12 +29,60 @@ jobs:
+      with:
+        name: ui-tests/ui-tests-${{ matrix.os }}-python${{ matrix.python }}
+        path: |
+            playwright-report
```
It seems it is not uploading artifacts due to path mismatch 

```suggestion
            ui-tests/playwright-report
```

krassowski commented on .github/workflows/tests.yml:
```diff
@@ -65,7 +65,7 @@ jobs:
       run: |
           cd ui-tests
-          npx playwright test
+          npx playwright test --update-snapshots
```
This should be undone; the test should indeed fail when there are no snapshots as otherwise we are not comparing against a reference.

HaudinFlorence replied:
This has been undone.

krassowski commented:
bot please update playwright snapshots

krassowski commented on .github/workflows/tests.yml:
```diff
@@ -28,12 +29,60 @@ jobs:
+      if: always()
+      uses: actions/upload-artifact@v2
+      with:
+        name: ui-tests/ui-tests-${{ matrix.os }}-python${{ matrix.python }}
```
```suggestion
        name: ui-tests-${{ matrix.os }}-python${{ matrix.python }}
```

Slashes in names not allowed as CI tells us

krassowski commented:
Hmm I do not see the actions picking up here - it might only start working after this PR is merged. You will get the snapshots in the report (after fixing the slash issue, see comment above) so that is probably enough to unblock it.

HaudinFlorence commented:
> Hmm I do not see the actions picking up here - it might only start working after this PR is merged. You will get the snapshots in the report (after fixing the slash issue, see comment above) so that is probably enough to unblock it.

I am not sure I properly adapted `update-integration-tests.yml`. I tried to run run the tetst in this PR https://github.com/HaudinFlorence/nbdime/actions/runs/6058727473/job/16441136493?pr=8

krassowski commented:
This is what I was referring to in https://github.com/jupyter/nbdime/pull/664#issuecomment-1703834598. I think it needs to be merged into main/master branch to kick in. Let's try on my fork: https://github.com/krassowski/nbdime/pull/4#issuecomment-1703898422

krassowski commented on .github/workflows/update-integration-tests.yml:
```diff
@@ -0,0 +1,53 @@
+      - name: Install extension
+        run: |
+          set -eux
+          npm
+          python -m pip install .
```
I would guess this needs something like:

```suggestion
          python -m pip install .
          npm run build
```

krassowski commented on .github/workflows/tests.yml:
```diff
@@ -28,12 +29,60 @@ jobs:
+      run: |
+          npm install -D @playwright/test
+          npx playwright install
+    - name: Install nbdime
+      run: |
+          pip install -e .[test]
+    - name: Install jupyterlab and dependencies
+      run: |
+          python -m pip install --upgrade pip wheel setuptools
+          python -m pip install jupyterlab~=4.0
```
I would swap these two steps (first install latest pip to ensure resolution only then install nbdime).

krassowski commented on .github/workflows/tests.yml:
```diff
@@ -28,12 +29,60 @@ jobs:
+          npx playwright install
+    - name: Install nbdime
+      run: |
+          pip install -e .[test]
```
Does it compile JS by itself or is `npm run build` needed to ensure we are testing the code on the branch rather than released?

HaudinFlorence replied:
I probably need to add `npm run build`.

krassowski commented on .github/workflows/tests.yml:
```diff
@@ -79,22 +128,20 @@ jobs:
             jupyter_server-version: '<2'
           - python-version: '3.11'
             jupyter_server-version: '>=2'
-          - python-version: '3.7'
-            os: windows-latest
```
I think this would be good to restore but bumping it to 3.8.

HaudinFlorence replied:
I am not sure to get what it means.

krassowski replied:
I mean restore the windows test at the lower end of the supported versions which now is 3.8.

krassowski replied:
My understanding is that the original intent here was: test on Linux across all versions and on Windows at the oldest and at the most recent. I think this is a good pattern in general.

krassowski replied:
See https://github.com/jupyter/nbdime/pull/664/files#r1314193757

krassowski commented on .github/workflows/tests.yml:
```diff
@@ -79,22 +127,20 @@ jobs:
-          - python-version: '3.7'
-            os: windows-latest
-          - python-version: '3.11'
+          - python-version: '3.8'
```
This is what I meant:

```suggestion
          - python-version: '3.8'
            os: windows-latest
          - python-version: '3.11'
            os: windows-latest
```

krassowski commented on .github/workflows/tests.yml:
```diff
@@ -28,12 +29,59 @@ jobs:
+      fail-fast: false
+      matrix:
+        os: [windows-latest, ubuntu-22.04]
+        python: ["3.7", "3.8", "3.9", "3.10", "3.11"]
```
```suggestion
        python: ["3.8", "3.9", "3.10", "3.11"]
```

krassowski replied:
I would even recommend the same pattern as in the `python` job where all Python versions are run on Linux but only oldest and latest is run on Windows.

krassowski commented:
Hi @HaudinFlorence please see https://github.com/HaudinFlorence/nbdime/pull/9.

krassowski commented on packages/nbdime/src/diff/widget/output.ts:
```diff
@@ -422,7 +421,7 @@ class RenderableOutputView extends RenderableDiffView<nbformat.IOutput> {
     let model = this.model;
     this.mimetype = mimeType;
-    each(this.layout.widgets, (w: Widget) => {
+    each(this.layout!.widgets, (w: Widget) => {
```
It should be possible to drop `each`, `find` and `toArray` in this file rewriting to native JS. We could then drop `@lumino/algorithm` here. This can be done after this PR is merged.

krassowski replied:
I would suggest adding this to PR description as a follow up task. I don't think this is worth changing now, there are more important things to address.

HaudinFlorence replied:
Ok. This has been added to the section "Future work and follow up tasks" of the description of this PR.

krassowski commented on packages/webapp/src/app/staticdiff.ts:
```diff
@@ -6,9 +6,9 @@ import {
 
-import {
+/* import {
   EditorWidget
-} from 'nbdime/lib/common/editor';
+} from 'nbdime/lib/common/editor'; */
```
Commented code to be removed

HaudinFlorence replied:
@krassowski Thanks. This will be removed.

HaudinFlorence commented:
> Hi @HaudinFlorence please see [HaudinFlorence#9](https://github.com/HaudinFlorence/nbdime/pull/9).

@krassowski Thanks a lot for your precious help on making these ui-tests work.

krassowski commented on packages/nbdime/src/styles/diff.css:
```diff
@@ -184,7 +194,7 @@
 .jp-CollapsiblePanel-container {
   background-color: var(--jp-layout-color1);
-  color: var(--jp-ui-font-color1);
+  color: var(--jp-ui--color1);
```
```suggestion
  color: var(--jp-ui-font-color1);
```

krassowski commented on packages/nbdime/test/src/diff/widget/metadata.spec.ts:
```diff
@@ -1,5 +1,6 @@
 // Copyright (c) Jupyter Development Team.
 // Distributed under the terms of the Modified BSD License.
+globalThis.DragEvent = class DragEvent {} as any;
```
Not needed because this was added to setup file. Sorry I missed it earlier.

```suggestion
```

krassowski commented on packages/nbdime/test/src/diff/widget/output.spec.ts:
```diff
@@ -1,6 +1,8 @@
 // Copyright (c) Jupyter Development Team.
 // Distributed under the terms of the Modified BSD License.
 
+globalThis.DragEvent = class DragEvent {} as any;
+
```
As above

```suggestion
```

krassowski reviewed (commented):
Hooray, checks are green :green_circle: :tada:

While the merge view appears fine from the snapshots, there are some problems with the diff view in JupyterLab (I have not checked the standalone apps):

1. There is an error in the console that Jeremy highlighted earlier:
   `Uncaught TypeError: can't access property "lineChunks", o is undefined`
2. The editor is not read-only. This is because CodeMirror6 uses a different settings system; more work will be needed to make the legacy options work here.
3. Added cells are not highlighted as such (maybe a CSS style got removed accidentally, or class name was not adjusted?)
4. Changed and added lines are highlighted in red
5. The `green` color used to highlight individual tokens is too strong making it difficult to read. Ideally the same colour as in previous version would be used.
6. The added cells take up too much space.
7. Line spacer is incorrectly placed. In the example below it is in the middle but should be at the end. Maybe it is off by one issue?

| previous version | this PR |
|---|---|
|![Screenshot from 2023-09-03 16-38-45](https://github.com/jupyter/nbdime/assets/5832902/e3d1673f-2efa-422f-b7c0-83de95a151ca) | ![Screenshot from 2023-09-03 16-38-15](https://github.com/jupyter/nbdime/assets/5832902/f7b75c76-c23b-4b3d-8e77-c39fc654b2b7) |

HaudinFlorence commented:
> Hooray, checks are green 🟢 🎉
> 
Great !
> While the merge view appears fine from the snapshots, there are some problems with the diff view in JupyterLab (I have not checked the standalone apps):


I must recognize that on the contrary I had only checked the apps and not the labextension.
> 
> 1. There is an error in the console that Jeremy highlighted earlier:
>    `Uncaught TypeError: can't access property "lineChunks", o is undefined`

I have got the same error if I am introducing a change in my notebook and then click directly on `git` . Sorry if it has no sense and if it is not the way the extension should be used but if we first save the modified notebook before clicking on the `git`button, then it seems to be ok and the diff result is displayed.

> 2. The editor is not read-only. This is because CodeMirror6 uses a different settings system; more work will be needed to make the legacy options work here.

> 3. Added cells are not highlighted as such (maybe a CSS style got removed accidentally, or class name was not adjusted?)

I will check this.
> 4. Changed and added lines are highlighted in red

I will check this too.
> 5. The `green` color used to highlight individual tokens is too strong making it difficult to read. Ideally the same colour as in previous version would be used.

I will try to change this.
> 6. The added cells take up too much space.

> 7. Line spacer is incorrectly placed. In the example below it is in the middle but should be at the end. Maybe it is off by one issue?

I had seen this issue, it is not supposed to happen at least in the app, because adding padding algorithm has a different condition to treat the case of last line of the document.

HaudinFlorence commented:
> 4. Changed and added lines are highlighted in red
This has been changed has shown on the capture inserted just after.

> 5. The `green` color used to highlight individual tokens is too strong making it difficult to read. Ideally the same colour as in previous version would be used.

Not sure the colours are exactly the same as in previous version. But hopefully it is better than before.
![comparisons_diff_greens](https://github.com/jupyter/nbdime/assets/99649086/94dc43c4-c530-4279-985b-f23036c6f7ad)

HaudinFlorence commented:
> 3. Added cells are not highlighted as such (maybe a CSS style got removed accidentally, or class name was not adjusted?)


> 6. The added cells take up too much space.

Points 3 and 6 should be fixed.
The following captures make a comparison between the rendering of the previous and of the update version.
![comparisons_added_cells](https://github.com/jupyter/nbdime/assets/99649086/2dd11ca6-9eb4-421d-a413-ae7e79333ce6)

HaudinFlorence commented:
I hope that point 7 is now fixed by commit saying "[Fix alignment shift for diff in alignViews.](https://github.com/jupyter/nbdime/pull/664/commits/25f395bfb7585552af498f151edc14153f9ec2e2)".

krassowski commented:
Thank you for all the fixes! With the latest commit the alignment looks good and the highlights in edited cells look great. One thing that does not work yet is the CSS for added/deleted cells - it does not show up for me, neither in JupyterLab nor in the standalone app:

![image](https://github.com/jupyter/nbdime/assets/5832902/6f49c2eb-10a3-49de-b463-349f3af492fa)

The editors of added/deleted cells have `cm-merge-pane-added`/`cm-merge-pane-deleted` class respectively but the "[Fix the added cells CSS style.](https://github.com/jupyter/nbdime/pull/664/commits/43e6b0190db0cd6e818266815e7a8c8d7bc8eef5)" commit does not modify styles for this class. What does work is the outputs of the added/deleted cells which are targeted by different set of selectors, like `.jp-Notebook-diff .jp-Diff-deleted .jp-Diff-renderedOutput`. It would be of great value if we had visual tests for a notebook where not only text is added and deleted but also cells with output are added and deleted.

Please let me know if there is anything I any help is needed with the remaining points (1) and (2).

HaudinFlorence commented:
> Thank you for all the fixes! With the latest commit the alignment looks good and the highlights in edited cells look great. One thing that does not work yet is the CSS for added/deleted cells - it does not show up for me, neither in JupyterLab nor in the standalone app:
> 
> ![image](https://user-images.githubusercontent.com/5832902/266038653-6f49c2eb-10a3-49de-b463-349f3af492fa.png)
> 
> The editors of added/deleted cells have `cm-merge-pane-added`/`cm-merge-pane-deleted` class respectively but the "[Fix the added cells CSS style.](https://github.com/jupyter/nbdime/pull/664/commits/43e6b0190db0cd6e818266815e7a8c8d7bc8eef5)" commit does not modify styles for this class. What does work is the outputs of the added/deleted cells which are targeted by different set of selectors, like `.jp-Notebook-diff .jp-Diff-deleted .jp-Diff-renderedOutput`. It would be of great value if we had visual tests for a notebook where not only text is added and deleted but also cells with output are added and deleted.
> 
> Please let me know if there is anything I any help is needed with the remaining points (1) and (2).

Thanks for your feedback @krassowski I need to check back for the cells background.

krassowski commented:
I think you need to go deeper with the selectors. This appears to be because `cm-editor` and `cm-gutters` have their own background occluding the cell background. You could make the cm-editor background transparent, or to acurately match how it looked in previous version:

![Screenshot from 2023-09-07 10-43-08](https://github.com/jupyter/nbdime/assets/5832902/6b795fb8-3842-43d9-85ad-26967a781849)


you could just target `cm-editor` like:


```css
.cm-merge-pane-added > .cm-editor {
   background: lightgreen;
   border: 1px solid green;
}
```

But using proper variables instead of colour names

HaudinFlorence commented:
@krassowski Thanks for your comment. Commit https://github.com/jupyter/nbdime/pull/664/commits/e59d8408652873521f5b7618e9b79b4497309601 should fix the issue with background of added/deleted cells

krassowski commented:
Once https://github.com/HaudinFlorence/nbdime/pull/11 is in I will approve this PR on the basis that the core functionality is working well and could be released as alpha or beta. There may be more bugs to fix, and certainly there are quite a few maintenance tasks:
- there are many examples added in this PR which are not used in tests, these need cleaning up/getting better names
- visual test need to cover the diff app as well, not only the merge app
- a prettier pass would really help (there is a PR for that already)
- we shall review API before final release.

krassowski reviewed (approved):
Thanks @HaudinFlorence.

What would make it easiest for you to move forward:
- wait for the UI tests and https://github.com/HaudinFlorence/nbdime/pull/12 to be merged into the branch and only then merge this PR
- merge this PR as-is and retarget the remaining PRs against master branch in this repo?

Review-wise this is getting difficult to handle due to conversation  length and number of files modified, but we can handle two more merges if that's easier.

HaudinFlorence commented:
> Thanks @HaudinFlorence.
> 
> What would make it easiest for you to move forward:
> 
> * wait for the UI tests and [A list of small improvements HaudinFlorence/nbdime#12](https://github.com/HaudinFlorence/nbdime/pull/12) to be merged into the branch and only then merge this PR
> * merge this PR as-is and retarget the remaining PRs against master branch in this repo?
> 
> Review-wise this is getting difficult to handle due to conversation length and number of files modified, but we can handle two more merges if that's easier.

Ok. I have tried to push the 3 added ui-tests but not sure this was a good idea to keep on moving forward before the tests are fixed on https://github.com/HaudinFlorence/nbdime/pull/12

HaudinFlorence commented:
bot please update playwright snapshots

vidartf commented:
As I mentioned previously, I've been traveling recently, so I've missed a lot of the discussion here. I will try to read through the discussion, but for now: I merged the lint PR, so now @HaudinFlorence is no longer a first-time contributor, so workflows should no longer need manual approvals.

vidartf commented:
> Thanks @HaudinFlorence.
> 
> What would make it easiest for you to move forward:
> 
>     * wait for the UI tests and [A list of small improvements HaudinFlorence/nbdime#12](https://github.com/HaudinFlorence/nbdime/pull/12) to be merged into the branch and only then merge this PR
> 
>     * merge this PR as-is and retarget the remaining PRs against master branch in this repo?
> 
> 
> Review-wise this is getting difficult to handle due to conversation length and number of files modified, but we can handle two more merges if that's easier.

If you want, I can create a new branch on this repo for now (~"lab4"). Will that help, or does the actions config need to be in the default branch?

krassowski commented:
It needs to be on the default branch I believe. I had created https://github.com/jupyter/nbdime/tree/3.2.x yesterday so I think we could merge this one into `master` (merge conflicts due to prettier apart).

HaudinFlorence commented:
a PR rebased with master is opened here https://github.com/jupyter/nbdime/pull/673

SylvainCorlay commented:
Closing as superseded by #673.
